### ISSUE TITLE
Dram Bandwidth below 20bytes/clk/channel for InterleavedToSharded

### ISSUE THREAD

**Issue by miacim** [tenstorrent/tt-metal#6729]
InterleavedToSharded operation has lower bandwidth when data format is BFLOAT8_B or BFLOAT4_B. Here is an example for tensor dimension used in MLP of Falcon7B.

```
             OP CODE  IN0_Y  IN0_X INPUT_0_DATA TYPE  DEVICE FW DURATION [ns]  i0_tile_size       bytes/clk
InterleavedToSharded   2048   4608           FLOAT32                   137986        4096             22
InterleavedToSharded   2048   4608          BFLOAT16                    74224        2048             21
InterleavedToSharded   1024   4608           FLOAT32                    73215        4096             21
InterleavedToSharded   2048   4608         BFLOAT8_B                    51907        1088             16
InterleavedToSharded   2048   4608         BFLOAT4_B                    43149         576             10
InterleavedToSharded   1024   4608          BFLOAT16                    40772        2048             19
InterleavedToSharded    512   4608           FLOAT32                    34342        4096             22
InterleavedToSharded   1024   4608         BFLOAT8_B                    27082        1088             15
InterleavedToSharded   1024   4608         BFLOAT4_B                    22265         576              9
InterleavedToSharded    512   4608          BFLOAT16                    18775        2048             20
InterleavedToSharded    256   4608           FLOAT32                    17208        4096             22
InterleavedToSharded    512   4608         BFLOAT8_B                    13555        1088             15
InterleavedToSharded    512   4608         BFLOAT4_B                    11685         576              9
InterleavedToSharded    256   4608          BFLOAT16                    10146        2048             19
InterleavedToSharded    256   4608         BFLOAT8_B                     7321        1088             14
InterleavedToSharded    256   4608         BFLOAT4_B                     6418         576              8
```

**Comment by jliangTT** [comment]
Hey @miacim , what is the performance impact of the overall model? try to prioritize this.

**Comment by jliangTT** [comment]
i will call this bug for now.. although i am not sure.

**Comment by davorchap** [comment]
@bbradelTT has started to look into optimizing these OPs

**Comment by bbradelTT** [comment]
@miacim could you please provide the steps that you used to generate the table?

**Comment by jliangTT** [comment]
sorry for the confusion. we are now using a generic op_cat: TM in place of op_cat: perf TM, op_cat:Pytorch TM, etc. I will delete these deprecated TM labels to avoid confusion.

**Comment by miacim** [comment]
> @miacim could you please provide the steps that you used to generate the table?

I used data generated from profile_this, and then just do the math. I tested on 1GHz clk frequency, so math is as follows, for first example.
```InterleavedToSharded   2048   4608           FLOAT32                   137986        4096             22```
Num_Channels = 12
Num_Tiles = 2048 * 4608 / 32 / 32
Num_Bytes = Num_Tiles * 4096
Bytes/Clk = (Num_Bytes / Fw_Duration) / Num_Channels = 22.8

**Comment by bbradelTT** [comment]
I looked at the code. I think there are a couple of reasons for the different number of bytes transferred.

First, the smaller data type sizes have extra data overhead on the tile.
tt_metal/common/constants.hpp shows:
constexpr uint32_t BFLOAT8_B_TILE_HW = TILE_HW + 64;        // Value is in bytes - 1B per BFP8 datum + 4x16B exponent section
constexpr uint32_t BFLOAT4_B_TILE_HW = (TILE_HW / 2) + 64;  // Value is in bytes - 0.5B per BFP4 datum + 4x16B exponent section
Effectively, bfloat8_b has a 7% overhead and bfloat4_b has a 13% overhead in terms of bytes per tile element.

Second, for tile layout, the code eventually reaches tt_eager/tt_dnn/op_library/sharded/kernels/dataflow/reader_unary_sharded_blocks_interleaved_start_id.cpp, which performs operations on a per tile basis, including calling a read barrier:
```
    for (uint32_t h = 0; h < block_height_tiles; h++) {
        uint32_t tile_id = curr_tile_id;
        for (uint32_t w = 0; w < block_width_tiles; w++) {
            noc_async_read_tile(tile_id, s, l1_write_addr);
            tile_id++;
            l1_write_addr += tile_bytes;
            noc_async_read_barrier();
        }
        curr_tile_id += input_width_offset_tiles;  
    }  
```
Therefore the smaller data type sizes have more overhead due to executing more code and having more read barriers per a given number of bytes. The overhead for that is much more than 5% (and probably above 10%) for the float32 case, and would be 2x, 4x, and 8x when the size goes down. Using 15%, and looking at the first few entries, the adjusted bytes/clk (bytes/clk*(extra byte mult + code/barrier mult) are pretty well aligned:

in0_y | in0_x | data type | bytes/clk | extra byte mult | code/barrier mult | adj. bytes/clk
-- | -- | -- | -- | -- | -- | --
2048 | 4608 | 32 | 22 | 0 | 1.15 | 25.3
2048 | 4608 | 16 | 21 | 0 | 1.3 | 27.3
2048 | 4608 | 8 | 16 | 0.07 | 1.6 | 26.72
2048 | 4608 | 4 | 10 | 0.13 | 2.2 | 23.3

I think we'd need to write more data at one time for a single read barrier and loop iteration to be able to make the code run faster.

That would mean either writing more tiles or somehow making the tiles bigger.

**Comment by davorchap** [comment]
> > @miacim could you please provide the steps that you used to generate the table?
> 
> I used data generated from profile_this, and then just do the math. I tested on 1GHz clk frequency, so math is as follows, for first example. `InterleavedToSharded 2048 4608 FLOAT32 137986 4096 22` Num_Channels = 12 Num_Tiles = 2048 * 4608 / 32 / 32 Num_Bytes = Num_Tiles * 4096 Bytes/Clk = (Num_Bytes / Fw_Duration) / Num_Channels = 22.8

Hey @miacim 

Is input in DRAM, if yes, then

Total memory BW achieved = total bytes / time 
(2048*4608/32/32) * 4096 B / 137986 cycles = 273 B/c,  this is basically 273 GB/s , which would be close to theoretical DRAM BW of 288 GB/s (IP spec).
 
Can you confirm?

**Comment by davorchap** [comment]
@yugaoTT it would be good to see if your u-bench with 4096 transfer size achieves the same BW

**Comment by miacim** [comment]
> > > @miacim could you please provide the steps that you used to generate the table?
> > 
> > 
> > I used data generated from profile_this, and then just do the math. I tested on 1GHz clk frequency, so math is as follows, for first example. `InterleavedToSharded 2048 4608 FLOAT32 137986 4096 22` Num_Channels = 12 Num_Tiles = 2048 * 4608 / 32 / 32 Num_Bytes = Num_Tiles * 4096 Bytes/Clk = (Num_Bytes / Fw_Duration) / Num_Channels = 22.8
> 
> Hey @miacim
> 
> Is input in DRAM, if yes, then
> 
> Total memory BW achieved = total bytes / time (2048*4608/32/32) * 4096 B / 137986 cycles = 273 B/c, this is basically 273 GB/s , which would be close to theoretical DRAM BW of 288 GB/s (IP spec).
> 
> Can you confirm?

Yes. Inputs are in DRAM. For Float32 and Float16 it is close to theoretical maximum.

**Comment by yugaoTT** [comment]
@davorchap I'm only getting ~185GB/s for 4096 tile size for the baseline test, and ~215GB/s for the hybrid NoC/round-robin VC test.

**Comment by yugaoTT** [comment]
@miacim I tried a interleaved to sharded test with input shape of 2048x4608 (FLOAT32), sharded on 8x8 cores, I get FW duration of 198611, which equals to 15.8 GB/s per channel. Not sure why we get diff perfs. Here is my test code

`
@pytest.mark.parametrize(
    "input_shape, shard_scheme, shard_size",
    [
        ([1, 1, 2048, 4608], ttl.tensor.TensorMemoryLayout.BLOCK_SHARDED, (256, 576)),
    ],
)
@pytest.mark.parametrize(
    "shard_orientation", [ttl.tensor.ShardOrientation.ROW_MAJOR]
)
@pytest.mark.parametrize("input_dtype", [ttl.tensor.DataType.FLOAT32])
@pytest.mark.parametrize("output_dtype", [ttl.tensor.DataType.FLOAT32])
def test_sharded_tile(
    device, input_shape, shard_size, shard_scheme, shard_orientation, input_dtype, output_dtype, function_level_defaults
):
    grid_size = device.compute_with_storage_grid_size()
    input_size = torch.Size(input_shape)

    x = torch.arange(input_size.numel()).reshape(input_size).bfloat16().float()

    xt = (
        ttl.tensor.Tensor(
            x.reshape(-1).tolist(),
            x.shape,
            input_dtype,
            ttl.tensor.Layout.ROW_MAJOR,
        )
        .to(ttl.tensor.Layout.TILE)
        .to(
            device,
            ttl.tensor.MemoryConfig(
                memory_layout=ttl.tensor.TensorMemoryLayout.INTERLEAVED,
                buffer_type=ttl.tensor.BufferType.DRAM,
            ),
        )
    )

    yt = ttl.tensor.interleaved_to_sharded(
        xt, grid_size, shard_size, shard_scheme, shard_orientation, output_dtype=output_dtype
    )

    zt = ttl.tensor.sharded_to_interleaved(
        yt,
        ttl.tensor.MemoryConfig(
            memory_layout=ttl.tensor.TensorMemoryLayout.INTERLEAVED,
            buffer_type=ttl.tensor.BufferType.DRAM,
        ),
        output_dtype=input_dtype,
    )

    tt_og = xt.cpu().to(ttl.tensor.Layout.ROW_MAJOR).to_torch()

    tt_got_back = zt.cpu().to(ttl.tensor.Layout.ROW_MAJOR).to_torch()

    if input_dtype == output_dtype:
        passing, output = comp_equal(tt_og, tt_got_back)
    else:
        passing, output = comp_pcc(tt_og, tt_got_back, 0.999)
    logger.info(output)

    assert passing
`

**Comment by yugaoTT** [comment]
@miacim could you please share your test for interleaved to sharded test with input shape of 2048x4608 (FLOAT32)
? I can try repro on my end.

**Comment by davorchap** [comment]
fyi @acejkov

**Comment by davorchap** [comment]
@miacim are you DRAM links on WH machine training to 12G ?

This is what my machine gets:
<img width="1718" alt="image" src="https://github.com/tenstorrent-metal/tt-metal/assets/3885633/a844a5d3-8c69-4d37-94bd-11a42382f91e">

**Comment by bbradelTT** [comment]
I tried calling noc_async_read_barrier() less frequently (e.g. moving it to the outer loop). That did not change the performance behaviour of my micro-benchmark. The overhead seems to be in the rest of the code.

Next, I'm looking at using both the reader and writer core as reader cores. However, the way that cb_reserve_back, get_write_ptr, and cb_push_back (in tt_metal/hw/inc/dataflow_api.h) work together, it looks like these cannot be called on multiple cores for the same circular buffer. Therefore I'm looking at using two circular buffers next.

**Comment by miacim** [comment]
It turned out that configuration of the test was L1 interleaved, not DRAM interleaved. Please check if there is something additionally unusual in code.

```python
@pytest.mark.parametrize(
    "dtype", [ttl.tensor.DataType.BFLOAT16, ttl.tensor.DataType.BFLOAT8_B,ttl.tensor.DataType.FLOAT32, ttl.tensor.DataType.BFLOAT4_B]
)
@pytest.mark.parametrize(
    "y", [256, 512, 1024, 2048])
def test_my_sharding(device, dtype, y):
    input_dtype = dtype
    output_dtype = dtype
    shard_scheme = ttl.tensor.TensorMemoryLayout.BLOCK_SHARDED
    shard_grid = ttl.tensor.CoreRangeSet({ttl.tensor.CoreRange(ttl.tensor.CoreCoord(0, 0), ttl.tensor.CoreCoord(7, 7))})
    
    interleaved_mem_config = ttl.tensor.MemoryConfig(
        memory_layout=ttl.tensor.TensorMemoryLayout.INTERLEAVED,
        buffer_type=ttl.tensor.BufferType.L1,
    )

    sharded_mem_config = ttl.tensor.MemoryConfig(
        memory_layout=ttl.tensor.TensorMemoryLayout.BLOCK_SHARDED,
        buffer_type=ttl.tensor.BufferType.L1,
    )
    x = torch.randn((1, 1, y, 4608)).bfloat16().float()

    xt = (
        ttl.tensor.Tensor(
            x.reshape(-1).tolist(),
            x.shape,
            input_dtype,
            ttl.tensor.Layout.ROW_MAJOR,
        )
        .to(ttl.tensor.Layout.TILE)
        .to(
            device,
            interleaved_mem_config,
        )
    )

    yt = ttl.tensor.interleaved_to_sharded(xt, shard_grid, (y//8, 576), shard_scheme, ttl.tensor.ShardOrientation.ROW_MAJOR )
```

**Comment by miacim** [comment]
Here are the numbers for interleaved DRAM with CLK 800MHz. There is still difference in performance for different data types.
```
                  OP CODE  IN0_Y  IN0_X INPUT_0_DATA TYPE  CORE COUNT  DEVICE FW DURATION [ns]  i0_tile_size  bytes/ns per channel
15   InterleavedToSharded   2048   4608           FLOAT32          64                   244274        4096.0                  12.0
11   InterleavedToSharded   1024   4608           FLOAT32          64                   130888        4096.0                  12.0
12   InterleavedToSharded   2048   4608          BFLOAT16          64                   121322        2048.0                  12.0
14   InterleavedToSharded   2048   4608         BFLOAT8_B          64                    85726        1088.0                   9.0
13   InterleavedToSharded   2048   4608         BFLOAT4_B          64                    73420         576.0                   6.0
7    InterleavedToSharded    512   4608           FLOAT32          64                    67160        4096.0                  11.0
8    InterleavedToSharded   1024   4608          BFLOAT16          64                    63189        2048.0                  12.0
10   InterleavedToSharded   1024   4608         BFLOAT8_B          64                    44682        1088.0                   9.0
3    InterleavedToSharded    256   4608           FLOAT32          64                    39072        4096.0                  10.0
9    InterleavedToSharded   1024   4608         BFLOAT4_B          64                    37932         576.0                   5.0
4    InterleavedToSharded    512   4608          BFLOAT16          64                    37370        2048.0                  10.0
6    InterleavedToSharded    512   4608         BFLOAT8_B          64                    24140        1088.0                   8.0
0    InterleavedToSharded    256   4608          BFLOAT16          64                    21186        2048.0                   9.0
5    InterleavedToSharded    512   4608         BFLOAT4_B          64                    20084         576.0                   5.0
2    InterleavedToSharded    256   4608         BFLOAT8_B          64                    14034        1088.0                   7.0
1    InterleavedToSharded    256   4608         BFLOAT4_B          64                    11304         576.0                   4.0
```

**Comment by pavlejosipovic** [comment]
I've implemented a heuristic to put a read barrier after 8k of data in read requests
https://github.com/tenstorrent-metal/tt-metal/tree/pjosipovic/6729
This has improved dram read 50% - Bfp4, 30% Bfp8, 10% Float16, and no change for Float32

**Comment by bbradelTT** [comment]
Thanks @pavlejosipovic

I need to figure out why I didn't see any performance change when I tried that.

One suggestion. It might be better for get_barrier_read_threshold() to be something like
```
uint32_t threshold = 2;
try {
    threshold = 8 / datum_size(data_format);
} catch (const std::exception &e) {
}
return threshold;
```
 There are other data types that can be 1, 2, or 4 bytes.

**Comment by pavlejosipovic** [comment]
I've measured this change on GS/WH, DRAM/L1 interleaved tensors and I've placed results here.
https://docs.google.com/spreadsheets/d/1uB3_AYX5PmID6DWW6WZaRN8D2FqTuTkp/edit#gid=1816091962
In general it seems that this change helps in all case.

**Comment by bbradelTT** [comment]
I talked to @pavlejosipovic 
I was looking at the wrong column.
He wants to extend the logic to matmuls and merge in one change.
